pkg/services/mssqlfg/database_pair_fe_primary_provision.go

```diff
@@ -0,0 +1,167 @@
+		return nil, fmt.Errorf("Secondary database with the name is already " +
+			"existed")
+	}
+	if err := validateFailoverGroup(
+		ctx,
+		&d.failoverGroupsClient,
+		ppp.GetString("primaryResourceGroup"),
+		pdt.PriServerName,
+		pp.GetString("failoverGroup"),
+	); err != nil {
+		if !strings.Contains(err.Error(), "ResourceNotFound") {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("Failover group with the name is already existed")
+	}
+	return instance.Details, nil
+}
+
+func (d *databasePairFePrimaryManager) getPriDatabase(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	pp := instance.ProvisioningParameters
+	ppp := instance.Parent.ProvisioningParameters
+	pdt := instance.Parent.Details.(*dbmsPairInstanceDetails)
+	if err := validateDatabase(
+		ctx,
+		&d.databasesClient,
+		ppp.GetString("primaryResourceGroup"),
+		pdt.PriServerName,
+		pp.GetString("database"),
+	); err != nil {
+		return nil, err
+	}
+	return instance.Details, nil
+}
+
+func (d *databasePairFePrimaryManager) deployPriARMTemplate(
+	_ context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	dt := instance.Details.(*databasePairInstanceDetails)
+	pdt := instance.Parent.Details.(*dbmsPairInstanceDetails)
+	pp := instance.ProvisioningParameters
+	ppp := instance.Parent.ProvisioningParameters
+	tagsObj := pp.GetObject("tags")
+	tags := make(map[string]string, len(tagsObj.Data))
+	for k := range tagsObj.Data {
+		tags[k] = tagsObj.GetString(k)
+	}
+	if err := deployDatabaseFeARMTemplate(
+		&d.armDeployer,
+		dt.PriARMDeploymentName,
+		ppp.GetString("primaryResourceGroup"),
+		ppp.GetString("primaryLocation"),
+		pdt.PriServerName,
+		pp.GetString("database"),
+		tags,
+	); err != nil {
+		return nil, fmt.Errorf("error deploying ARM template: %s", err)
+	}
+	return instance.Details, nil
+}
+
+func (d *databasePairFePrimaryManager) deployFgARMTemplate(
+	_ context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	pp := instance.ProvisioningParameters
+	if err := deployFailoverGroupARMTemplate(
+		&d.armDeployer,
+		instance,
+	); err != nil {
+		return nil, fmt.Errorf("error deploying ARM template: %s", err)
+	}
+	dt := instance.Details.(*databasePairInstanceDetails)
+	dt.FailoverGroupName = pp.GetString("failoverGroup")
+	return dt, nil
+}
+
+func (d *databasePairFePrimaryManager) deploySecARMTemplate(
+	_ context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	dt := instance.Details.(*databasePairInstanceDetails)
+	pdt := instance.Parent.Details.(*dbmsPairInstanceDetails)
+	pp := instance.ProvisioningParameters
+	ppp := instance.Parent.ProvisioningParameters
+	pd := instance.Plan.GetProperties().Extended["tierDetails"].(planDetails)
+	tagsObj := pp.GetObject("tags")
+	tags := make(map[string]string, len(tagsObj.Data))
+	for k := range tagsObj.Data {
+		tags[k] = tagsObj.GetString(k)
+	}
+	if err := deployDatabaseARMTemplate(
```

Comment:
Will previous step `deployFgARMTemplate` automatically create a secondary database?

Reply:
That's right. And the secondary database must be created by the creation of failover group. Or it would fail to create the failover group. BTW, The deployment for the secondary database is for update api only. When the SKU of the primary database is changed, the secondary won't change automatically. The broker should keep their SKUs in sync.

Reply:
If no new database will be created here, using function `databaseFeARMTemplateBytes` should be more exact. And seems it won't affect the update step. In update step, `updateDatabaseARMTemplate` will be invoked, no conflict here...?

Reply:
Good catch. Thanks. Will change it to `Fe`.